pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -63,108 +65,113 @@ func NewWatcher(cfg *config.PipedSpec, gitClient gitClient, logger *zap.Logger)
-		}
+		w.providerCfgs[cfg.Name] = cfg
+	}
 
+	for _, repo := range w.config.Repositories {
 		w.wg.Add(1)
-		go w.run(ctx, p, cfg.PullInterval.Duration())
+		go w.run(ctx, &repo)
 	}
+
 	w.wg.Wait()
 	return nil
 }
 
-// run periodically compares the image in the given provider and one in git repository.
+// run periodically compares the image in the given git repository and one in the image provider.
 // And then pushes those with differences.
-func (w *watcher) run(ctx context.Context, provider imageprovider.Provider, interval time.Duration) {
+func (w *watcher) run(ctx context.Context, repoCfg *config.PipedRepository) {
 	defer w.wg.Done()
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
 
+	var (
+		pullInterval = defaultPullInterval
+		commitMsg    = ""
+		includedCfgs = []string{}
+		excludedCfgs = []string{}
+	)
+	// Use user-defined settings if there is.
+	for _, r := range w.config.ImageWatcher.Repos {
+		if r.RepoID != repoCfg.RepoID {
+			continue
+		}
+		pullInterval = time.Duration(r.PullInterval)
+		commitMsg = r.CommitMessage
+		includedCfgs = r.Includes
+		excludedCfgs = r.Excludes
+		break
+	}
+
+	// Periodically update this cloned directory as long as this worker continues.
+	repo, err := w.gitClient.Clone(ctx, repoCfg.RepoID, repoCfg.Remote, repoCfg.Branch, "")
+	if err != nil {
+		w.logger.Error("failed to clone repository",
+			zap.String("repo-id", repoCfg.RepoID),
+			zap.Error(err),
+		)
+		return
+	}
+
+	ticker := time.NewTicker(pullInterval)
+	defer ticker.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			// Inspect all targets in all repos, and compare only images managed by the given provider.
-			for id, repo := range w.gitRepos {
-				cfg, err := w.loadImageWatcherConfig(ctx, id, repo)
-				if err != nil {
-					w.logger.Error("failed to load image watcher config",
-						zap.String("repo-id", id),
+			err := repo.Pull(ctx, repo.GetClonedBranch())
+			if err != nil {
+				w.logger.Error("failed to perform git pull",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.String("branch", repo.GetClonedBranch()),
+					zap.Error(err),
+				)
+				continue
+			}
+			cfg, ok, err := config.LoadImageWatcher(repo.GetPath(), includedCfgs, excludedCfgs)
+			if err != nil {
+				w.logger.Error("failed to load configuration file for Image Watcher",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+				continue
+			}
+			if !ok {
+				w.logger.Error("configuration file for Image Watcher not found",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+				continue
+			}
+			// Inspect all targets defined in the repo, and update outdated images.
+			for _, target := range cfg.Targets {
+				if err := w.updateOutdatedImage(ctx, &target, repoCfg, repo.GetPath(), commitMsg); err != nil {
+					w.logger.Error("failed to update image",
+						zap.String("repo-id", repoCfg.RepoID),
```

Comment:
Instead of checking and updating for each target, how about checking and adding the commit for each target and then pushing once after all.

Reply:
Thank you. That way can reduce the number of pushing to git. But I'm a bit worried about the commit will be unclear. Including all changes into a single commit may make it tough to revert

Reply:
I know this is mainly for dev environment, but just warried.

Reply:
No, we add multiple commits (each commit for each change) and push after all.

Reply:
Ah, now I can understand. For that, we should share a temporary directory between all targets, but it looks like no problem to do.

Reply:
Thank you! It looks the best way!

Reply:
👍